packages/rum-core/src/common/truncate.js

```diff
@@ -0,0 +1,239 @@
+/**
+ * MIT License
+ *
+ * Copyright (c) 2017-present, Elasticsearch BV
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
+/**
+ * Goes over the object in Breadth first fashion
+ * Example:
+ * {
+ *    a: '10',
+ *    b: { c: 20, d: { e: 30 } },
+ *    f: { g: '100' }
+ * }
+ *
+ * Execute callbacks on below paths in order
+ * ['a'],  ['b', 'c'], ['b', 'd', 'e'], ['f', 'g']
+ */
+
+function breathFilter(obj = {}, callback, path = []) {
+  const queue = []
+  let index = 0
+  const keys = Object.keys(obj)
+  if (keys.length === index) {
+    return obj
+  }
+  queue.push(keys[index])
+
+  while (queue.length > 0) {
+    const currKey = queue.shift()
+    const value = obj[currKey]
+
+    if (typeof value === 'object') {
+      breathFilter(value, callback, path.concat(currKey))
+    } else {
+      obj[currKey] = callback(value, path.concat(currKey))
+    }
+    index++
+    if (index < keys.length) {
+      queue.push(keys[index])
+    }
+  }
+  return obj
+}
+
+function truncate(value, limit, required = false, placeholder = 'N/A') {
+  if (required && !value) {
+    value = placeholder
+  }
+  if (value) {
+    return String(value).substr(0, limit)
+  }
+  return value
+}
+
+function truncateMetadata(metadata, opts) {
+  return breathFilter(metadata, (value, path) => {
+    if (typeof value !== 'string') {
+      return value
+    }
+    let limit
+    let required = false
+    switch (path[0]) {
+      case 'service': {
+        switch (path[1]) {
+          case 'version':
+          case 'environment':
+            limit = opts.stringLimit
```

Comment:
We don't differentiate between those in our codebase. Can you give an example? I checked each and every field in server spec and most of them is 1024. Thats why i went with the limit. And we dont have something like sql or db query like backend agents..

Reply:
For example custom context fields are not keywords so they don't have 1024 limit on them but we still need to limit them at a higher number to make sure the payload has predictable range.  #46 is actually only about non-keyword fields since keyword fields are already being limited. In other words, even if a field doesn't have 1024 limit on the APM server spec we still need to limit them on the client but at a higher limit maybe (10000 characters),

Reply:
Hmmm, I got your concern, but currently we dont limit lot of other fields at 1024. so this PR takes care of that as well. But as for the non-keyword fields. I am fine with introducing an upper limit as per the suggestion.

Reply:
We should not close #46 if we're not limiting non-keyword fields, I'm ok with doing it in another PR though. Please remove #46 from the description of this PR in that case.